Allow the restore operation to mark a backup as preserved

kmichel-aiven opened the pull request:
## About this change - What it does
A preserved backup won't be automatically deleted even if it matches
the automatic deletion rules.

The backups following the preserved backup won't be deleted either.

This capability is enabled by passing the `--preserve-until` flag
during restoration.

Multiple preservation requests can co-exist and be separately
canceled for the same backup.

The effective preservation duration is the longest of all the
uncancelled preservation requests.

If the storage is not writable and the option was enabled, pghoard
will emit an error message but will not fail the restoration.

The preservation request is canceled at the end of the restore
operation, only if the restoration is successful..

## Why this way

The request is not renewed or extended during the backup restoration.
The expectation is that the option is used with a far enough date is
the future that it's not a problem.

Using a far-future date is unlikely to cause issues during normal
operations since the request is canceled after a successful restore.

If needed, an operator could write a preservation request in the
storage from outside pghoard, pghoard will see it and honour it

kmichel-aiven committed:
Allow the restore operation to mark a backup as preserved

A preserved backup won't be automatically deleted even if it matches
the automatic deletion rules.

The backups following the preserved backup won't be deleted either.

This capability is enabled by passing the `--preserve-until` flag
during restoration.

Multiple preservation requests can co-exist and be separately
canceled for the same backup.

The effective preservation duration is the longest of all the
uncancelled preservation requests.

If the storage is not writable and the option was enabled, pghoard
will emit an error message but will not fail the restoration.

The preservation request is canceled at the end of the restore
operation, only if the restoration is successful.

The request is not renewed or extended during the backup restoration.
The expectation is that the option is used with a far enough date is
the future that it's not a problem.

Using a far-future date is unlikely to cause issues during normal
operations since the request is canceled after a successful restore.

If needed, an operator could write a preservation request in the
storage from outside pghoard, pghoard will see it and honour it.

kmichel-aiven committed:
Split ObjectStore and related out of restore.py

pylint was rightfully complaining about the excessive line count.

alanfranz commented on test/test_restore.py:
```diff
@@ -575,3 +578,29 @@ def test_restore_get_delta_basebackup_data():
     assert all(f in files for f in expected_files)
+
+
+def test_object_store_request_backup_preservation(tmp_path: Path) -> None:
```
Wouldn't we like an additional kind of integration test with
- request backup preservation
- perform backup deletion according to the various rules
- verify the backup is there (or, even better, than it can be restored)

I mean, I see a lot of (good) unit testing, but maybe I'm missing an overall test that gives me guarantees around the preservation behaviour.

kmichel-aiven replied:
Good point, I added tests for that. It's far from perfect, but it spans most of the code involved in backup/preservation/deletion.

alanfranz reviewed (commented):
I think this is mostly ok (maybe one test is missing or I don't understand where I should look for it). Tests are failing though.

kmichel-aiven committed:
test: add backup preservation integration test

It's not a perfect test: our setup is not well suited to testing
concurrent operations (a backup in progress and a deletion). Fixing
that is more than just adding threads, we need ways to stop between
step for instance.

To cover for the lack of concurrent execution, there is a new option
to not cancel preservation at the end of a successful restore.
It could also be useful in real scenarios, if the restoration needs
some external validation before marking it as successful.

kmichel-aiven committed:
test: test backup preservation failures

Make sure write errors fail without killing the entire backup process.

alanfranz reviewed (approved):
LGTM
